File: domain/src/main/java/nam/tran/domain/Repository.java
package nam.tran.domain;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.Transformations;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.util.List;

import javax.inject.Inject;

import nam.tran.domain.entity.SoccerSeasonEntity;
import nam.tran.domain.entity.TeamEntity;
import nam.tran.domain.entity.state.Resource;
import nam.tran.domain.executor.AppExecutors;
import nam.tran.domain.interactor.core.DataBoundResource;
import nam.tran.domain.mapper.DataEntityMapper;
import nam.tran.flatform.IApi;
import nam.tran.flatform.core.ApiResponse;
import nam.tran.flatform.database.DbProvider;
import nam.tran.flatform.local.IPreference;
import nam.tran.flatform.model.response.session.SoccerSeason;
import nam.tran.flatform.model.response.team.ListTeamResponse;
import nam.tran.flatform.model.response.team.Team;

import static nam.tran.domain.entity.state.Loading.LOADING_NORMAL;

@SuppressWarnings({"FieldCanBeLocal", "unused"})
public class Repository implements IRepository {

    private final AppExecutors appExecutors;
    private final IPreference iPreference;
    private final DataEntityMapper dataEntityMapper;
    private final IApi iApi;
    private final DbProvider dbProvider;

    @Inject
    Repository(AppExecutors appExecutors, IPreference iPreference, DataEntityMapper dataEntityMapper, IApi iApi, DbProvider dbProvider) {
        this.appExecutors = appExecutors;
        this.iPreference = iPreference;
        this.dataEntityMapper = dataEntityMapper;
        this.iApi = iApi;
        this.dbProvider = dbProvider;
    }

    @Override
    public LiveData<Resource<List<SoccerSeasonEntity>>> getListSeason() {
        return new DataBoundResource<List<SoccerSeasonEntity>, List<SoccerSeason>>(appExecutors) {
            @Override
            protected void saveCallResult(@NonNull List<SoccerSeason> item) {
                dbProvider.beginTransaction();
                try {
                    dbProvider.soccerSeasonDao().insert(item);
                    dbProvider.setTransactionSuccessful();
                } finally {
                    dbProvider.endTransaction();
                }
            }

            @Override
            protected boolean shouldFetch(@Nullable List<SoccerSeasonEntity> data) {
                return data == null || data.isEmpty();
            }

            @NonNull
            @Override
            protected LiveData<List<SoccerSeasonEntity>> loadFromDb() {
                return Transformations.switchMap(dbProvider.soccerSeasonDao().fetchSoccerSeason(), object -> {
                    MutableLiveData<List<SoccerSeasonEntity>> data = new MutableLiveData<>();
                    data.setValue(dataEntityMapper.mSoccerSeasonEntityDataMapper.transform(object));
                    return data;
                });
            }

            @NonNull
            @Override
            protected LiveData<ApiResponse<List<SoccerSeason>>> createCall() {
                return iApi.getListSeason();
            }

            @Override
            protected int statusLoading() {
                return LOADING_NORMAL;
            }
        }.asLiveData();
    }

    @Override
    public LiveData<Resource<List<TeamEntity>>> getListTeam(int idSeason, int type) {
        return new DataBoundResource<List<TeamEntity>, ListTeamResponse>(appExecutors) {
            @Override
            protected void saveCallResult(@NonNull ListTeamResponse item) {
                for (Team team : item.getTeams()){
                    team.idSeason = idSeason;
                }
                dbProvider.beginTransaction();
                try {
                    dbProvider.teamDao().insert(item.getTeams());
                    dbProvider.setTransactionSuccessful();
                } finally {
                    dbProvider.endTransaction();
                }
            }

            @Override
            protected boolean shouldFetch(@Nullable List<TeamEntity> data) {
                return data == null || data.isEmpty();
            }

            @NonNull
            @Override
            protected LiveData<List<TeamEntity>> loadFromDb() {
                return Transformations.switchMap(dbProvider.teamDao().fetchTeams(idSeason), input -> {
                    MutableLiveData<List<TeamEntity>> data = new MutableLiveData<>();
                    data.setValue(dataEntityMapper.mTeamEntityDataMapper.transform(input));
                    return data;
                });
            }

            @NonNull
            @Override
            protected LiveData<ApiResponse<ListTeamResponse>> createCall() {
                return iApi.getListTeam(idSeason);
            }

            @Override
            protected int statusLoading() {
                return type;
            }
        }.asLiveData();
    }
}

File: domain/src/main/java/nam/tran/domain/Repository.kt
package nam.tran.domain

import android.arch.lifecycle.LiveData
import android.arch.lifecycle.MutableLiveData
import android.arch.lifecycle.Transformations
import nam.tran.domain.entity.SoccerSeasonEntity
import nam.tran.domain.entity.TeamEntity
import nam.tran.domain.entity.state.Loading.LOADING_NORMAL
import nam.tran.domain.entity.state.Resource
import nam.tran.domain.executor.AppExecutors
import nam.tran.domain.interactor.core.DataBoundResource
import nam.tran.domain.mapper.DataEntityMapper
import nam.tran.flatform.IApi
import nam.tran.flatform.core.ApiResponse
import nam.tran.flatform.database.DbProvider
import nam.tran.flatform.local.IPreference
import nam.tran.flatform.model.response.session.SoccerSeason
import nam.tran.flatform.model.response.team.ListTeamResponse
import javax.inject.Inject

@Suppress("unused")
class Repository @Inject
internal constructor(private val appExecutors: AppExecutors, private val iPreference: IPreference, private val dataEntityMapper: DataEntityMapper, private val iApi: IApi, private val dbProvider: DbProvider) : IRepository {

    override val listSeason: LiveData<Resource<List<SoccerSeasonEntity>>>
        get() = object : DataBoundResource<List<SoccerSeasonEntity>, List<SoccerSeason>>(appExecutors) {
            override fun saveCallResult(item: List<SoccerSeason>) {
                dbProvider.beginTransaction()
                try {
                    dbProvider.soccerSeasonDao().insert(item)
                    dbProvider.setTransactionSuccessful()
                } finally {
                    dbProvider.endTransaction()
                }
            }

            override fun shouldFetch(data: List<SoccerSeasonEntity>?): Boolean {
                return data == null || data.isEmpty()
            }

            override fun loadFromDb(): LiveData<List<SoccerSeasonEntity>> {
                return Transformations.switchMap(dbProvider.soccerSeasonDao().fetchSoccerSeason()) { `object` ->
                    val data = MutableLiveData<List<SoccerSeasonEntity>>()
                    data.value = dataEntityMapper.mSoccerSeasonEntityDataMapper.transform(`object`)
                    data
                }
            }

            override fun createCall(): LiveData<ApiResponse<List<SoccerSeason>>> {
                return iApi.listSeason
            }

            override fun statusLoading(): Int {
                return LOADING_NORMAL
            }
        }.asLiveData()

    override fun getListTeam(idSeason: Int, type: Int): LiveData<Resource<List<TeamEntity>>> {
        return object : DataBoundResource<List<TeamEntity>, ListTeamResponse>(appExecutors) {
            override fun saveCallResult(item: ListTeamResponse) {
                for (team in item.teams!!) {
                    team.idSeason = idSeason
                }
                dbProvider.beginTransaction()
                try {
                    dbProvider.teamDao().insert(item.teams!!)
                    dbProvider.setTransactionSuccessful()
                } finally {
                    dbProvider.endTransaction()
                }
            }

            override fun shouldFetch(data: List<TeamEntity>?): Boolean {
                return data == null || data.isEmpty()
            }

            override fun loadFromDb(): LiveData<List<TeamEntity>> {
                return Transformations.switchMap(dbProvider.teamDao().fetchTeams(idSeason)) { input ->
                    val data = MutableLiveData<List<TeamEntity>>()
                    data.value = dataEntityMapper.mTeamEntityDataMapper.transform(input)
                    data
                }
            }

            override fun createCall(): LiveData<ApiResponse<ListTeamResponse>> {
                return iApi.getListTeam(idSeason)
            }

            override fun statusLoading(): Int {
                return type
            }
        }.asLiveData()
    }
}
